modules/geochemistry/src/userobjects/NodalVoidVolume.C

```diff
@@ -0,0 +1,226 @@
+        if (_nodal_void_volume.count(elem->node_ptr(i)) == 0)
+          _nodal_void_volume[elem->node_ptr(i)] = 0.0;
+  if (_app.n_processors() > 1)
+    buildCommLists();
+  _rebuilding_needed = false;
+}
+
+void
+NodalVoidVolume::buildCommLists()
+{
+  // global_node_nums_to_receive[pid] is a list of node numbers connected to elements owned by
+  // processor pid.  _nodes_to_receive[pid] is a list of node pointers.  Both lists have the same
+  // ordering
+  std::map<processor_id_type, std::vector<dof_id_type>> global_node_nums_to_receive;
+  _nodes_to_receive.clear();
+  // run through all elements known by this processor (the owned + 1 layer of ghosted elements),
+  // recording nodes that are attached to elements that aren't owned by this processor
+  for (const auto & elem : _fe_problem.getEvaluableElementRange())
+    if (this->hasBlocks(elem->subdomain_id()))
+    {
+      const processor_id_type elem_pid = elem->processor_id();
+      if (elem_pid != _my_pid)
+      {
+        if (global_node_nums_to_receive.count(elem_pid) == 0)
+        {
+          global_node_nums_to_receive[elem_pid] = std::vector<dof_id_type>();
+          _nodes_to_receive[elem_pid] = std::vector<const Node *>();
+        }
+        for (unsigned i = 0; i < elem->n_nodes(); ++i)
+          if (std::find(global_node_nums_to_receive[elem_pid].begin(),
+                        global_node_nums_to_receive[elem_pid].end(),
+                        elem->node_id(i)) == global_node_nums_to_receive[elem_pid].end())
+          {
+            global_node_nums_to_receive[elem_pid].push_back(elem->node_id(i));
+            _nodes_to_receive[elem_pid].push_back(elem->node_ptr(i));
+          }
+      }
+    }
+
+  // exchange this info with other processors, building global_node_nums_to_send at the same time
+  std::map<processor_id_type, std::vector<dof_id_type>> global_node_nums_to_send;
+  auto nodes_action_functor = [&](processor_id_type pid, const std::vector<dof_id_type> & nts) {
+    global_node_nums_to_send[pid] = nts;
+  };
+  Parallel::push_parallel_vector_data(
+      this->comm(), global_node_nums_to_receive, nodes_action_functor);
+
+  // Build _nodes_to_send using global_node_nums_to_send, keeping the same ordering in the
+  // std::vector
+  _nodes_to_send.clear();
+  for (const auto & kv : global_node_nums_to_send)
+  {
+    const processor_id_type pid = kv.first;
+    _nodes_to_send[pid] = std::vector<const Node *>();
+    for (const auto & node_num : kv.second)
+      _nodes_to_send[pid].push_back(_mesh.nodePtr(node_num));
+  }
+}
+
+void
+NodalVoidVolume::exchangeGhostedInfo()
+{
+  // Exchange ghosted _nodal_void_volume information with other processors
+  std::map<processor_id_type, std::vector<Real>> nvv_to_send;
+  for (const auto & kv : _nodes_to_send)
+  {
+    const processor_id_type pid = kv.first;
+    nvv_to_send[pid] = std::vector<Real>();
+    for (const auto & nd : kv.second)
+      nvv_to_send[pid].push_back(_nodal_void_volume.at(nd));
```

Comment:
```suggestion
    auto & pid_entry = nvv_to_send[pid];
    pid_entry.reserve(kv.second.size());
    for (const auto & nd : kv.second)
      pid_entry.push_back(_nodal_void_volume.at(nd));
```
Same here. Avoids multiple map lookups and reserves.

Reply:
Thanks